src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -26,103 +26,71 @@
+        int size = (data != null) ? data.limit() : 0;
+        int len = 4;
+        len += fastIntLength(size);
+        if (replyTo != null) {
+            if (utf8mode) {
+                len += fastUtf8Length(replyTo) + 1;
             } else {
-                idx--;
-                lengthBytes[idx] = digits[0];
-            }
-
-            // Build the array
-            int len = 4 + subject.length() + 1 + (lengthBytes.length - idx);
-
-            if (replyTo != null) {
                 len += replyTo.length() + 1;
             }
+        }
+        if (utf8mode) {
+            len += fastUtf8Length(subject) + 1;
+            charset = StandardCharsets.UTF_8;
+        } else {
+            len += subject.length() + 1;
+            charset = StandardCharsets.US_ASCII;
+        }
+        this.protocolBytes = ByteBuffer.allocate(len);
+        protocolBytes.put((byte)'P').put((byte)'U').put((byte)'B').put((byte)' ');
+        protocolBytes.put(subject.getBytes(charset));
+        protocolBytes.put((byte)' ');
 
-            this.protocolBytes = new byte[len];
-
-            // Copy everything
-            int pos = 0;
-            protocolBytes[0] = 'P';
-            protocolBytes[1] = 'U';
-            protocolBytes[2] = 'B';
-            protocolBytes[3] = ' ';
-            pos = 4;
-            pos = copy(protocolBytes, pos, subject);
-            protocolBytes[pos] = ' ';
-            pos++;
-
-            if (replyTo != null) {
-                pos = copy(protocolBytes, pos, replyTo);
-                protocolBytes[pos] = ' ';
-                pos++;
-            }
+        if (replyTo != null) {
+            protocolBytes.put(replyTo.getBytes(charset));
+            protocolBytes.put((byte)' ');
+        }
 
-            System.arraycopy(lengthBytes, idx, protocolBytes, pos, lengthBytes.length - idx);
+        if (size > 0) {
+            int base = protocolBytes.limit();
+            for (int i = size; i > 0; i /= 10) {
+                base--;
+                protocolBytes.put(base, (byte)(i % 10 + 0x30));
+            }
+        } else {
+            protocolBytes.put((byte)0x30);
         }
+        protocolBytes.clear();
+    }
 
-        this.sizeInBytes = this.protocolBytes.length + data.length + 4;// for 2x \r\n
+    NatsMessage(String subject, String replyTo, byte[] data, boolean utf8mode) {
+        this(subject, replyTo, ByteBuffer.wrap(data), utf8mode);
     }
 
     // Create a protocol only message to publish
     NatsMessage(CharBuffer protocol) {
-        ByteBuffer byteBuffer = StandardCharsets.UTF_8.encode(protocol);
-        this.protocolBytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
-        Arrays.fill(byteBuffer.array(), (byte) 0); // clear sensitive data
-        this.sizeInBytes = this.protocolBytes.length + 2;// for \r\n
+        protocol.mark();
+        this.protocolBytes = ByteBuffer.allocate(fastUtf8Length(protocol));

```

Comment:
By pre-allocating a correctly sized `ByteBuffer` we can eliminate the slow path that involves copying bytes in `getProtocolBytes()`.